gax/call_context.h

```diff
@@ -0,0 +1,150 @@
+ * the correct ClientContext.
+ *
+ * Example usage, with both customization and rpc invocation:
+ *
+ * @code
+ * gax::Status OuterStub::GetFoo(gax::CallContext& context,
+ *                               GetFooRequest const& request,
+ *                               Foo* response) {
+ *   int counter = 0;
+ *   context.AddGrpcContextPolicy([&counter, &context](grpc::ClientContext* c){
+ *       counter++;
+ *       c->set_cacheable(true);
+ *       c->set_idempotent(context.Info().idempotency ==
+ *                         google::gax::MethodInfo::Idempotency::IDEMPOTENT);
+ *     });
+ *   context.AddMetadata("fooKey", "fooVal");
+ *   // Assume retry_stub_ engages in some kind of retry, holds a ptr to an
+ *   // InnerStub, and calls inner_stub_->GetFoo.
+ *   gax::Status status = retry_stub_->GetFoo(context, request, response);
+ *   std::cout << "Needed " << counter << " attempts for GetFoo" << std::endl;
+ *   return status;
+ * }
+ *
+ * gax::Status InnerStub::GetFoo(gax::CallContext& context,
+ *                               GetFooRequest const& request,
+ *                               Foo* response) {
+ *   grpc::ClientContext c;
+ *   context.PrepareGrpcContext(&c);
+ *   return grpc_stub_->GetFoo(&c, request, response);
+ * }
+ */
+class CallContext {
+ public:
+  CallContext(MethodInfo method_info)
+      : deadline_(std::chrono::time_point<std::chrono::system_clock>::max()),
+        method_info_(std::move(method_info)) {}
+
+  /**
+   * Register an arbitrary customization function on grpc::ClientContext.
+   * This function could tweak advanced knobs or provide other custom behavior.
+   */
+  void AddGrpcContextPolicy(GrpcContextPolicyFunc f);
+
+  /**
+   * Initializes a grpc::ClientContext, adding stored metadata, setting a
+   * deadline, and applying registered customization functions.
+   *
+   * @par Pre-conditions
+   * The 'context' parameter is pristine and has not been used for an rpc.
+   *
+   * @param context the context to initialize.
+   */
+  void PrepareGrpcContext(grpc::ClientContext* context);
+
+  /**
+   * @name Register application-specific metadata.
+   *
+   * @param key the metadata key.
+   * @param val the metadata value.
+   *
+   * @return True if the key had not previously been registered, false
+   * otherwise.
+   */
+  bool AddMetadata(std::string key, std::string val);
+
+  /**
+   * @name Set a deadline for the rpc.
+   */
+  void SetDeadline(std::chrono::system_clock::time_point deadline);
+
+  /**
+   * @name Accessor for configured rpc deadline.
+   */
+  std::chrono::system_clock::time_point Deadline() const;
+
+  /**
+   * @name Accessor for method info.
+   */
+  MethodInfo Info() const { return method_info_; }
+
+ private:
+  FRIEND_TEST(CallContext, Basic);
+  std::chrono::system_clock::time_point deadline_;
+  std::vector<GrpcContextPolicyFunc> context_policies_;
+  std::map<std::string const, std::string const> metadata_;
```

Comment:
This needs to be a multimap - it is valid to register the same key multiple times.

Reply:
Done